### PR TITLE
[FIRRTL] LowerTypes + ExpandWhens: fixes for high firrtl

### PR DESCRIPTION
**Pull request by youngar**
There are 4 unrelated changes in this PR that I was too lazy to split apart.

[FIRRTL][LowerTypes] Extend SubaccessOp hack
    
    The subaccess op is not supported yet in the lower-types pass. To get
    around this limitation, there is an incorrect lowering of this
    operation.  The subaccess op is lowered to always return the first index
    of the array.
    
    This change extends the hack to use support full lowering when there is
    a vector of bundles.  It does this by creating a Subindex[0] op, and
    lowering that.
    
    This also change the diagnostic from an `error` to a `warning`.

[FIRRTL][ExpandWhens] Make sure values are passive before muxing
    
    When a sink is used as a the RHS of a connect statement, the expand
    whens pass may create a mux using that value.  The connect op can handle
    operands with flip type, while the mux op cannot.  This checks if the
    type is a flip and emits a the asPassive cast if it is not.

[FIRRTL][ExpandWhens] Fix type of created muxes
    
    The return type of muxes needs to be inferred.  In the case where the
    two operands are not the same type, it should be inferred to be the
    larger of the two.

[FIRRTL][LowerTypes] Convert partialconnects to connects
    
    This modifies the lower tyes pass to canonicalize all partial connects
    as regular connects.  This is something that happens in the SFC duing
    the expand-connects pass, which is a separate earlier pass from
    lower-types.
    
    A partial connect has implicit truncation when the source value is
    larger than the destination. When lower-types converts this to a regular
    connect, it will emit a tail-op to truncate the source first.
    
    Partial connects are partially common because the FIRRTL we are
    consuming has implicit truncations in regular connect statements, which
    we transform in to partial connects.
    
    Removal of partial connects and in particular implicit truncation needs
    to happen before the expand-whens pass runs.  The ExpandWhens pass will
    try to combine multiple connects to a destination into a single connect
    statement by muxing together the sources.  It would be tricky to have
    expand whens properly truncate all source values to create the correct
    mux operation.

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp [hunk 1]
```diff
@@ -152,16 +152,23 @@ class ExpandWhensVisitor : public FIRRTLVisitor<ExpandWhensVisitor> {
         condition.getLoc(), condition.getType(), condition, value);
   }
 
+  /// If a value has an outer flip, convert the value to passive.
+  Value convertToPassive(OpBuilder &builder, Location loc, Value input) {
+    auto inType = input.getType().cast<FIRRTLType>();
+    if (inType.isPassive())
+      return input;
+    return builder.create<AsPassivePrimOp>(loc, inType.getPassiveType(), input);
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp:160]: createAndFold<AsPassivePrimOp> will handle the "already passive" case for you

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp [hunk 2]
```diff
@@ -846,14 +849,13 @@ void TypeLoweringVisitor::visitExpr(SubfieldOp op) {
 
 // Gracefully die on subaccess operations
 void TypeLoweringVisitor::visitExpr(SubaccessOp op) {
-  op.emitError("SubaccessOp not handled.");
+  emitWarning(op.getLoc(), "SubaccessOp not handled.");
```
Comment (lattner) [inline lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:852]: Is turning this error into a warning the right thing?

**Review by lattner — APPROVED** [review]

File: lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp (hunk 2 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp]: I dropped this commit from the PR, it was not a good idea

File: lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp (hunk 1 above)
Comment (youngar) [reply lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp]: Thanks!

**Review by seldridge — APPROVED** [review]
Looks great!

I'm annoyed about the need for the `convertToPassive`. This should be far less likely to occur after https://github.com/llvm/circt/pull/1262 _for parsed FIRRTL only_. Annoyingly, the IR that the `convertToPassive` is guarding against should be required until we can change memories and instances to store directionality not using flips.

Procedurally, it would make sense to me to keep the history here and either rebase and merge through the GitHub UI or land these individually on `main`.

**Pull request merged**